File: C/0004.c
#include <stdio.h>

double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size) {
    // Ensure that nums1 is the smaller array to minimize the binary search range
    if (nums1Size > nums2Size) {
        return findMedianSortedArrays(nums2, nums2Size, nums1, nums1Size);
    }
    
    int m = nums1Size;
    int n = nums2Size;
    int halfLen = (m + n + 1) / 2;
    int imin = 0, imax = m;
    
    while (imin <= imax) {
        int i = (imin + imax) / 2;
        int j = halfLen - i;
        
        if (i < m && nums1[i] < nums2[j - 1]) {
            imin = i + 1; // i is too small, must increase it
        } else if (i > 0 && nums1[i - 1] > nums2[j]) {
            imax = i - 1; // i is too large, must decrease it
        } else {
            // i is perfect
            int max_of_left;
            if (i == 0) {
                max_of_left = nums2[j - 1];
            } else if (j == 0) {
                max_of_left = nums1[i - 1];
            } else {
                max_of_left = (nums1[i - 1] > nums2[j - 1]) ? nums1[i - 1] : nums2[j - 1];
            }
            
            if ((m + n) % 2 == 1) {
                return max_of_left;
            }
            
            int min_of_right;
            if (i == m) {
                min_of_right = nums2[j];
            } else if (j == n) {
                min_of_right = nums1[i];
            } else {
                min_of_right = (nums1[i] < nums2[j]) ? nums1[i] : nums2[j];
            }
            
            return (max_of_left + min_of_right) / 2.0;
        }
    }
    
    return 0.0; // Should never reach here if the input arrays are valid
}

File: Kotlin/0004.kt
class Solution {
    fun findMedianSortedArrays(nums1: IntArray, nums2: IntArray): Double {
        if (nums1.size > nums2.size) {
            // Ensure nums1 is the smaller array
            return findMedianSortedArrays(nums2, nums1)
        }
        
        val m = nums1.size
        val n = nums2.size
        var iMin = 0
        var iMax = m
        val halfLen = (m + n + 1) / 2
        
        while (iMin <= iMax) {
            val i = (iMin + iMax) / 2
            val j = halfLen - i
            
            val nums1LeftMax = if (i == 0) Int.MIN_VALUE else nums1[i - 1]
            val nums1RightMin = if (i == m) Int.MAX_VALUE else nums1[i]
            val nums2LeftMax = if (j == 0) Int.MIN_VALUE else nums2[j - 1]
            val nums2RightMin = if (j == n) Int.MAX_VALUE else nums2[j]
            
            if (nums1LeftMax <= nums2RightMin && nums2LeftMax <= nums1RightMin) {
                // Perfect partition
                return if ((m + n) % 2 == 0) {
                    (maxOf(nums1LeftMax, nums2LeftMax).toDouble() + minOf(nums1RightMin, nums2RightMin).toDouble()) / 2
                } else {
                    maxOf(nums1LeftMax, nums2LeftMax).toDouble()
                }
            } else if (nums1LeftMax > nums2RightMin) {
                // Move `i` to the left
                iMax = i - 1
            } else {
                // Move `i` to the right
                iMin = i + 1
            }
        }
        
        throw IllegalArgumentException("Input arrays are not sorted properly.")
    }
}
